RecoHGCal/TICL/plugins/SeedingRegionByL1.h

```diff
@@ -0,0 +1,43 @@
+// Author: Swagata Mukherjee
+
+#ifndef RecoHGCal_TICL_SeedingRegionByL1_h
+#define RecoHGCal_TICL_SeedingRegionByL1_h
+#include <memory>
+#include <string>
+
+#include "DataFormats/L1TCorrelator/interface/TkEm.h"
+#include "DataFormats/L1TCorrelator/interface/TkEmFwd.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "RecoHGCal/TICL/plugins/SeedingRegionAlgoBase.h"
+
+namespace ticl {
+  class SeedingRegionByL1 final : public SeedingRegionAlgoBase {
+  public:
+    SeedingRegionByL1(const edm::ParameterSet& conf, edm::ConsumesCollector& sumes);
+
+    void initialize(const edm::EventSetup& es) override{};
+    void makeRegions(const edm::Event& ev, const edm::EventSetup& es, std::vector<TICLSeedingRegion>& result) override;
+    static void fillPSetDescription(edm::ParameterSetDescription& desc);
+    static edm::ParameterSetDescription makePSetDescription();
```

Comment:
Please, see my comments [here](https://github.com/cms-sw/cmssw/pull/33001#discussion_r584639918).

Reply:
Hello @rovere,
just trying to understand it better,
do you mean something like this https://github.com/cms-sw/cmssw/blob/20cffd182934015d0fa68d8ddda953725b5a2e1f/RecoHGCal/TICL/plugins/TICLSeedingRegionProducer.cc#L69-L75 ?
i.e, the parameters needed by SeedingRegionByL1 class are included explicitly in the TICLSeedingRegionProducer::fillDescriptions ?

Reply:
Ciao @swagata87 I meant following the example of what has been done, e.g, here:

* base class for all algos [link](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h)
* derived plugins with proper implementation of `fillPSetDescription` [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h#L88) and [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/plugins/HGCalCLUEAlgo.h#L86)
* Create a `PluginFactory` and properly register the plugins as having also a `PSet` available: [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterPluginFactory.cc#L8-L11)
* Get the desired plugin from the `PluginFactory` by name: [here](https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HGCalRecProducers/plugins/HGCalLayerClusterProducer.cc#L88-L95)

All that is described in more details [here](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He)

Reply:
Ciao @Sam-Harper 
if we assume you also would like to have a backport, I agree with your plan 1, provided there's a commitment to also implement 2 after 1 has been integrated. This will have interplay with #33001 that could be used, in principle, as the PR where this approach is first implemented (eventually integrating also the changes required in these files).

Reply:
for me is ok you proceed as in 1. (i.e. merge the PR as is), and them I adjust all these Pset/plugins in my HFnose PR with possibly your comments and help (I do not think we need to backport mine)

Reply:
thank you, we are discussing the best way in trigger to proceed (doing this change will break config compatibility as it will move the track seeding parameters) to a "seeding" pset. So we are trying to figure out the best solution. 

Swagatas class I think should work "as is" as a plugin given it has a fillPSetDescription function which looks compliant.

Reply:
so thank you both again. We've had a discussion and we think we might want the new style config for 11_1 as well sooner rather than later to ensure we keep config compatibility with 11_3_X.  So we'll look into making these changes  in this PR and if they are simple, we'll just implement them and we can all discuss that implementation. If it looks like it wont converge, we'll just go for option 1 for speed reasons.

Reply:
so we updated the config, there are few things to address:

the name of plugin pset, we chose "seedingPSet"

the name of the parameter which is the plugin name, "type" was the example but its not great to use a python reserved word (although in this case we dont shadow). They can be changed to anything people want.

Reply:
@rovere any comments on the parameter names ?

Reply:
The name is fine.
Thanks @Sam-Harper @swagata87  for this additional effort!

Reply:
no problem it ended up being easier than we thought it would be at first glance :)

Reply:
> The name is fine.
> Thanks @Sam-Harper @swagata87 for this additional effort!

Thank you @rovere for your suggestion to use plugins, and also for the examples & twiki that you linked to before. 
The additional effort is made by Sam, he implemented the plugin related changes to address your comment. I wasn't fully sure how to do it quickly enough :-)